nemo_aligner/models/nlp/gpt/megatron_gpt_ppo_actor.py

```diff
@@ -67,80 +84,97 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
-                "attention_mask": attention_mask,
-                "position_ids": position_ids,
-                "advantages": advantages,
-                "prev_log_probs": prev_logprobs,
-                "mask": mask,
-            }
-
             required_keys = set()
             if parallel_state.get_pipeline_model_parallel_world_size() == 1:
                 required_keys.update(batch.keys())
             else:
                 required_keys.add("attention_mask")
 
                 if parallel_state.is_pipeline_first_stage():
-                    required_keys.update(("tokens", "position_ids"))
+                    required_keys.update(("response_tokens", "position_ids"))
 
                 if parallel_state.is_pipeline_last_stage():
-                    required_keys.update(("tokens", "advantages", "mask", "prev_log_probs"))
+                    required_keys.update(("response_tokens", "advantages", "mask", "prev_logprobs", "is_end"))
 
             batch = {key: val.cuda(non_blocking=True) if key in required_keys else None for key, val in batch.items()}
 
-            parallel_logits = model(batch["tokens"], batch["position_ids"], batch["attention_mask"], labels=None,)
+            parallel_logits = model(
+                batch["response_tokens"], batch["position_ids"], batch["attention_mask"], labels=None,
+            )
 
             def loss_func(parallel_logits):
                 mask = batch["mask"]
                 advantages = batch["advantages"]
-                prev_log_probs = batch["prev_log_probs"]
-                tokens = batch["tokens"]
+                prev_log_probs = batch["prev_logprobs"]
+                tokens = batch["response_tokens"]
+                is_end = batch["is_end"]
+
+                is_end_mask = mask * is_end.view(-1, 1)
 
-                curr_log_probs = from_parallel_logits_to_logprobs(vocab_parallel_logits=parallel_logits, target=tokens)
+                curr_log_probs = from_parallel_logits_to_logprobs(
+                    vocab_parallel_logits=parallel_logits, target=tokens, higher_stability=True
+                )
 
                 scaled_entropy = torch.tensor(0.0, dtype=parallel_logits.dtype, device=parallel_logits.device)
                 if self.entropy_bonus > 0:
-                    scaled_entropy = calculate_distributed_entropy(parallel_logits, mask) * self.entropy_bonus
+                    scaled_entropy = calculate_distributed_entropy(parallel_logits, is_end_mask) * self.entropy_bonus
 
                 # Calculate clipped PPO surrogate loss function.
                 ratios = (curr_log_probs - prev_log_probs).exp()
                 ratios_clamped = ratios.clamp(1.0 - self.ratio_eps, 1.0 + self.ratio_eps)
 
                 loss1 = -advantages * ratios
                 loss2 = -advantages * ratios_clamped
-                actor_loss = masked_mean(torch.max(loss1, loss2), mask)
-                loss = actor_loss - scaled_entropy
+
+                if is_end_mask.sum() > 0:
```

Comment:
I still want to see these metrics when nothing ended properly. otherwise i have to set these to 0 if i don't compute it when is_end_mask.sum() == 0 or something arbitrary which might be also confusing

Reply:
I see, so just to be sure having NaNs in `ppo_ratio` or `ppo_ratio_clamped` isn't going to cause issues?

Reply:
it should be ok. I don't think i've seen it nan ever since we enabled higher stability, also ppo_ratio and ppo_ratio_clamped doesn't use the is_end_mask so there is no change to this logic from the mask

Reply:
Ok sounds good